File: LegacySource/settlement/room/main/RoomBlueprintImp.cs
using System;
using System.Collections.Generic;
using System.IO;
using game.audio;
using game.boosting;
using game.faction;
using init.sprite;
using init.type;
using init.value;
using settlement.path.finders;
using settlement.room.main.category;
using settlement.room.main.furnisher;
using settlement.room.main.util;
using snake2d.util.color;
using snake2d.util.misc;
using util.info;

public abstract class RoomBlueprintImp : RoomBlueprint
{
    public readonly Icon icon;
    public readonly string type;
    public readonly INFO info;
    public readonly RoomCategorySub cat;
    protected double degradeRate;
    private readonly RoomUpgrades upgrades;
    private readonly int typeIndex;
    public readonly Lockable<Faction> reqs;

    public readonly Ambiance soundAmbiance;
    public readonly SoundRace clickSounds;

    protected Boostable bonus;

    static List<RoomBlueprintImp> IMPS = new List<RoomBlueprintImp>();
    static RoomBlueprintImp()
    {
        GameDisposable.Add(() =>
        {
            IMPS.Clear();
        });
    }

    protected RoomBlueprintImp(RoomInitData init, int typeIndex, string key, RoomCategorySub cat) : this(init, typeIndex, key, cat, null)
    {
    }

    protected RoomBlueprintImp(RoomInitData init, int typeIndex, string key, RoomCategorySub cat, ACTION wiki) : base(key)
    {
        init.Init(key);
        this.typeIndex = typeIndex;
        if (cat != null)
            cat.Add(this);
        this.type = init.Type();
        info = new INFO(init.Text(), wiki);
        icon = Icon(init);
        this.cat = cat;
        degradeRate = init.Data().Has("DEGRADE_RATE") ? init.Data().D("DEGRADE_RATE", 0, 1) : 0.75;
        upgrades = new RoomUpgrades(this, init);
        IMPS.Add(this);
        reqs = GVALUES.FACTION.LOCK.Push("ROOM_" + key, info.Name, info.Desc, icon);
        reqs.Push(init.Data());
        soundAmbiance = AUDIO.AMBI().Get("ROOM_" + key);
        clickSounds = AUDIO.Race("ROOM_CLICK_" + key);
    }

    private Icon Icon(RoomInitData init)
    {
        return SPRITES.icons().Get(init.Data());
    }

    public Boostable Bonus()
    {
        return bonus;
    }

    public Icon IconBig()
    {
        return icon;
    }

    public abstract SFinderFindable Service(int tx, int ty);

    public override COLOR MiniC(int tx, int ty)
    {
        return Constructor().MiniColor(tx, ty);
    }

    public override COLOR MiniCPimped(ColorImp original, int tx, int ty, bool northern, bool southern)
    {
        return Constructor().MiniColorPimped(original, tx, ty, northern, southern);
    }

    public abstract Furnisher Constructor();

    public bool IsAvailable(CLIMATE c)
    {
        return true;
    }

    public double DegradeRate()
    {
        return degradeRate;
    }

    public RoomUpgrades Upgrades()
    {
        return upgrades;
    }

    public int TypeIndex()
    {
        return typeIndex;
    }

    public override string ToString()
    {
        return "[" + Index() + "]" + Key;
    }
}

File: settlement/room/main/RoomBlueprintImp.java
package settlement.room.main;

import java.io.IOException;

import game.GameDisposable;
import game.audio.AUDIO;
import game.audio.Ambiance;
import game.audio.SoundRace;
import game.boosting.Boostable;
import game.faction.Faction;
import init.sprite.SPRITES;
import init.sprite.UI.Icon;
import init.type.CLIMATE;
import init.value.GVALUES;
import init.value.Lockable;
import settlement.path.finders.SFinderFindable;
import settlement.room.main.category.RoomCategorySub;
import settlement.room.main.furnisher.Furnisher;
import settlement.room.main.util.RoomInitData;
import snake2d.util.color.COLOR;
import snake2d.util.color.ColorImp;
import snake2d.util.misc.ACTION;
import snake2d.util.sets.ArrayListResize;
import util.info.INFO;

public abstract class RoomBlueprintImp extends RoomBlueprint{

	public final Icon icon;
	public final String type;
	public final INFO info;
	public final RoomCategorySub cat;
	protected double degradeRate;
	private final RoomUpgrades upgrades;
	private final int typeIndex;
	public final Lockable<Faction> reqs;
	
	public final Ambiance soundAmbiance;
	public final SoundRace clickSounds;
	
	protected Boostable bonus;

	static ArrayListResize<RoomBlueprintImp> IMPS = new ArrayListResize<>(10, 512);
	static {
		new GameDisposable() {
			@Override
			protected void dispose() {
				IMPS.clear();
			}
		};
	}
	
	protected RoomBlueprintImp(RoomInitData init, int typeIndex, String key, RoomCategorySub cat) throws IOException{
		this(init, typeIndex, key, cat, null);
		
	}

	protected RoomBlueprintImp(RoomInitData init, int typeIndex, String key, RoomCategorySub cat, ACTION wiki) throws IOException{
		super(key);
		init.init(key);
		this.typeIndex = typeIndex;
		if (cat != null)
			cat.add(this);
		this.type = init.type();
		info = new INFO(init.text(), wiki);
		icon = icon(init);
		this.cat = cat;
		if (init.data().has("DEGRADE_RATE"))
			degradeRate = init.data().d("DEGRADE_RATE", 0, 1);
		else
			degradeRate = 0.75;
		upgrades = new RoomUpgrades(this, init);
		IMPS.add(this);		
		reqs = GVALUES.FACTION.LOCK.push("ROOM_" + key, info.name, info.desc, icon);
		reqs.push(init.data());
		soundAmbiance = AUDIO.AMBI().get("ROOM_" + key);
		clickSounds = AUDIO.race("ROOM_CLICK_" + key);
	}
	
	private Icon icon(RoomInitData init) throws IOException{
		return SPRITES.icons().get(init.data());
	}
	

	
	public final Boostable bonus(){
		return bonus;
	}
	
	public Icon iconBig(){
		return icon;
	}

	@Override
	public abstract SFinderFindable service(int tx, int ty);

	
	@Override
	public COLOR miniC(int tx, int ty) {
		return constructor().miniColor(tx, ty);
	}
	
	@Override
	public COLOR miniCPimped(ColorImp origional, int tx, int ty, boolean northern, boolean southern) {
		return constructor().miniColorPimped(origional, tx, ty, northern, southern);
	}
	
	public abstract Furnisher constructor();
	
	
	public boolean isAvailable(CLIMATE c) {
		return true;
	}
	

	
	public double degradeRate() {
		return degradeRate;
	}
	
	public RoomUpgrades upgrades() {
		return upgrades;
	}
	
	public int typeIndex() {
		return typeIndex;
	}
	
	@Override
	public String toString() {
		return "["+index()+"]" + key;
	}
	
}
